bindings/java/unicorn_Unicorn.c

```diff
@@ -28,768 +28,1605 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
-         break;
-      case UC_HOOK_MEM_FETCH_UNMAPPED:    // Hook for all invalid memory access events
-      case UC_HOOK_MEM_READ_UNMAPPED:    // Hook for all invalid memory access events
-      case UC_HOOK_MEM_WRITE_UNMAPPED:    // Hook for all invalid memory access events
-      case UC_HOOK_MEM_FETCH_PROT:    // Hook for all invalid memory access events
-      case UC_HOOK_MEM_READ_PROT:    // Hook for all invalid memory access events
-      case UC_HOOK_MEM_WRITE_PROT:    // Hook for all invalid memory access events
-         if (invokeEventMemCallbacks == 0) {
-            invokeEventMemCallbacks = (*env)->GetStaticMethodID(env, clz, "invokeEventMemCallbacks", "(JIJIJ)Z");
-         }
-         err = uc_hook_add((uc_engine*)eng, &hh, (uc_hook_type)type, cb_eventmem, env, 1, 0);
-         break;
-   }
-   return (jlong)hh;
-}
-
-/*
- * Class:     unicorn_Unicorn
- * Method:    registerHook
- * Signature: (JII)J
+JNIEXPORT jlong JNICALL
+Java_unicorn_Unicorn__1hook_1add__JILunicorn_Hook_2Ljava_lang_Object_2JJ(
+    JNIEnv *env, jobject self, jlong uc, jint type, jobject callback,
+    jobject user_data, jlong begin, jlong end)
+{
+    const char *hook_sig;
+    void *hook_callback;
+
+    if (type == UC_HOOK_INTR) {
+        hook_sig = sig_InterruptHook;
+        hook_callback = cb_hookintr;
+    } else if (type == UC_HOOK_CODE || type == UC_HOOK_BLOCK) {
+        hook_sig = sig_CodeHook; // also BlockHook
+        hook_callback = cb_hookcode;
+    } else if ((type & UC_HOOK_MEM_INVALID) && !(type & ~UC_HOOK_MEM_INVALID)) {
+        hook_sig = sig_EventMemHook;
+        hook_callback = cb_eventmem;
+    } else if ((type & UC_HOOK_MEM_VALID) && !(type & ~UC_HOOK_MEM_VALID)) {
+        hook_sig = sig_MemHook;
+        hook_callback = cb_hookmem;
+    } else if (type == UC_HOOK_INSN_INVALID) {
+        hook_sig = sig_InvalidInstructionHook;
+        hook_callback = cb_hookinsn_invalid;
+    } else if (type == UC_HOOK_EDGE_GENERATED) {
+        hook_sig = sig_EdgeGeneratedHook;
+        hook_callback = cb_edge_gen;
+    } else if (type == UC_HOOK_TLB_FILL) {
+        hook_sig = sig_TlbFillHook;
+        hook_callback = cb_tlbevent;
+    } else {
+        throwUnicornException(env, UC_ERR_HOOK);
+        return 0;
+    }
+
+    struct hook_wrapper *hh =
+        makeHookWrapper(env, self, callback, user_data, "hook", hook_sig);
+    if (hh == NULL) {
+        return 0;
+    }
+    uc_err err = uc_hook_add((uc_engine *)uc, &hh->uc_hh, type, hook_callback,
+                             hh, begin, end);
+    if (err != UC_ERR_OK) {
+        throwUnicornException(env, err);
+        deleteHookWrapper(env, hh);
+        return 0;
+    }
+    return (jlong)hh;
+}
+
+/*
+ * Class:     unicorn_Unicorn
+ * Method:    _hook_add
+ * Signature: (JILunicorn/Hook;Ljava/lang/Object;JJI)J
+ */
+JNIEXPORT jlong JNICALL
+Java_unicorn_Unicorn__1hook_1add__JILunicorn_Hook_2Ljava_lang_Object_2JJI(
```

Comment:
Why this signatures looks so weird?

Reply:
`_hook_add` is overloaded: there are zero, one and two-argument variants (for hooks like `UC_HOOK_INSN` and `UC_HOOK_TCG_OPCODE` which take extra arguments). The JNI method name needs to be mangled in this case. The names and the comment are all auto-generated by  `javac -h`.

Reply:
Could you also document a bit (for examples, commands to generate this file etc) on how these files are generated? Or do I miss something?

The ultimate goal is that if some code is auto-generated, it should be reproducible.

Reply:
It's already part of the Makefile:

```
unicorn_Unicorn.h: unicorn/Unicorn.java
	$(JC) -h . $<
```

I can add a comment to the top of `unicorn_Unicorn.c` reminding people to keep the definitions in sync with the .h file.